### PR TITLE
feat: add gstreamer wrapper

### LINKED ISSUE #441: add gstreamer wrapper
a gstreamer wrapper with support of main loop, pipeline, appsrc, and dynamic element creation is needed.

### PR DESCRIPTION
**Pull request by cthulhu-irl**
# Pull Request Checklist

Please check if your pull request fulfills the following requirements:

  * [x] The commit message follows our guidelines
  * [x] Tests for the changes have been added / updated (for feat / fix / refactor)
  * [x] Docs for the changes have been added / updated (for feat / fix / refactor)

## Pull Request Type

What kind of change does this pull request introduce?

  * [ ] Build: Changes that affect the build system or external dependencies
  * [ ] CD: Changes to our CD configuration files and scripts
  * [ ] CI: Changes to our CI configuration files and scripts
  * [ ] Docs: Documentation only changes
  * [x] Feat: A new feature
  * [ ] Fix: A bug fix
  * [ ] Perf: A code change that improves performance
  * [ ] Refactor: A code change that neither fixes a bug nor adds a feature
  * [ ] Test: Adding missing tests or correcting existing tests

## What is the current behavior?

Issue #441

## What is the new behavior?

A gstreamer wrapper library within media module has been added.
 - internal wrapping, access, and ownership base components and utils.
 - core types, signal handler, and other helper classes.
 - common elements, including appsrc.
 - video/audio info and caps.
 - pipeline and mainloop abstractions.

## Does this Pull Request introduce a breaking change?

  * [ ] Yes
  * [x] No

## Other information

Closes #441

### CODE REVIEW

File: wolf/media/gst/audio/w_audio_info.hpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+#include "media/gst/internal/w_wrapper.hpp"
+#include "media/gst/core/w_caps.hpp"
+#include "media/gst/audio/w_audio_format.hpp"
+
+#include <gst/audio/audio-info.h>
+
+#include <stdexcept>
+#include <utility>
+
+namespace wolf::gst {
```
Comment (PooyaEimandar) [inline wolf/media/gst/audio/w_audio_info.hpp:15]: please change it to wolf::media::gst

File: wolf/media/gst/core/w_caps.hpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+#include <gst/gst.h>
+
+#include <string_view>
+
+namespace wolf::gst {
+
+/**
+ * @brief wrapper of GstCaps, gstreamer capabilities concept.
+ */
+class w_caps : public w_wrapper<w_caps, GstCaps, void, gst_caps_unref>
```
Comment (PooyaEimandar) [inline wolf/media/gst/core/w_caps.hpp:18]: please avoid object-oriented programming style in WOLF.

File: wolf/media/gst/core/w_element.hpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+namespace wolf::gst {
+
+/**
+ * @brief wrapper of GstElement, gstreamer elements process stream data.
+ *
+ * elements combine to make dataflow pipeline graphs (DAG)
+ * to process one or more media stream(s).
+ */
+class w_element : public w_wrapper<w_element, GstElement, void, gst_object_unref>
```
Comment (PooyaEimandar) [inline wolf/media/gst/core/w_element.hpp:22]: same here

File: wolf/media/test/gstreamer.hpp [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            });
+
+            BOOST_LEAF_AUTO(test_video_src,
+                    gst::w_element_factory::make_simple("videotestsrc", "src"));
+            BOOST_LEAF_AUTO(video_convert,
+                    gst::w_element_factory::make_simple("videoconvert", "video_convert"));
+            BOOST_LEAF_AUTO(auto_video_sink,
+                    gst::w_element_factory::make_simple("autovideosink", "sink"));
+
+            auto flow = gst::w_flow_path::make(
```
Comment (PooyaEimandar) [inline wolf/media/test/gstreamer.hpp:102]: Can we add another pipeline via string command?  
e.g: we want to append something like this to the end of your pipeline.
"queue ! ffenc_mpeg4 bitrate=800000 ! ffmux_mp4 ! filesink location=video_test.mp4"

**Review by PooyaEimandar — CHANGES_REQUESTED** [review]

File: wolf/media/test/gstreamer.hpp (hunk 4 above)
Comment (cthulhu-irl) [reply wolf/media/test/gstreamer.hpp:102]: yes, the `make_from_launch_str` factory function is for this purpose.

File: wolf/media/gst/core/w_caps.hpp (hunk 2 above)
Comment (cthulhu-irl) [reply wolf/media/gst/core/w_caps.hpp:18]: I found it inevitable unfortunately. this is a CRTP kind of inheritance, although `w_wrapper` doesn't have much to do with the derived (`Template` part in `CRTP`), but it is indeed a repeating pattern better lifted into `w_wrapper`.

It provides a common interface expressed via `internal::RawAccess`, keeping different wrappers interconnected.

I couldn't find another design without contravening DRY and SRP principles, resulting in too many and hard to keep sync/maintained repeated code.

File: wolf/media/gst/audio/w_audio_info.hpp (hunk 1 above)
Comment (cthulhu-irl) [reply wolf/media/gst/audio/w_audio_info.hpp]: done at https://github.com/WolfEngine/WolfEngine/pull/452/commits/e15b76708175917332bb40fbd92cc7b3f6965d17.

**Review by PooyaEimandar — APPROVED** [review]

**Comment by PooyaEimandar** [comment]
Thank you
@cthulhu-irl

**Pull request merged**